lib/auth/clt.go

```diff
@@ -74,48 +56,103 @@ const (
 	MissingNamespaceError = "missing required parameter: namespace"
 )
 
-// Dialer defines dialer function
-type Dialer func(network, addr string) (net.Conn, error)
+// ContextDialer type alias for backwards compatibility
+type ContextDialer = client.ContextDialer
+
+// ContextDialerFunc type alias for backwards compatibility
+type ContextDialerFunc = client.ContextDialerFunc
 
-// Client is HTTP Auth API client. It works by connecting to auth servers
-// via HTTP.
+// APIClient is aliased here so that it can be embedded in Client
+type APIClient = client.Client
+
+// Client is the Auth API client. It works by connecting to auth servers
+// via gRPC and HTTP.
 //
 // When Teleport servers connect to auth API, they usually establish an SSH
 // tunnel first, and then do HTTP-over-SSH. This client is wrapped by auth.TunClient
 // in lib/auth/tun.go
 type Client struct {
+	// APIClient is embedded so that Client can inherit its grpc endpoint
+	// methods to satisfy the ClientI interface. Client uses APIClient.Config
+	// for its own config, except TLS config is kept separate.
+	APIClient
+	// http client is deprecated and will be gradually phased out in favor of APIClient (gRPC).
 	sync.Mutex
-	ClientConfig
 	roundtrip.Client
-	transport  *http.Transport
-	conn       *grpc.ClientConn
-	grpcClient proto.AuthServiceClient
-	// closedFlag is set to indicate that the services are closed
-	closedFlag int32
+	transport *http.Transport
+	// TLS holds the TLS config for the http client.
+	TLS *tls.Config
 }
 
 // Make sure Client implements all the necessary methods.
 var _ ClientI = &Client{}
 
-// TLSConfig returns TLS config used by the client, could return nil
-// if the client is not using TLS
-func (c *Client) TLSConfig() *tls.Config {
-	return c.ClientConfig.TLS
-}
+// NewClient returns a new client that uses mutual TLS authentication
+// and dials the remote server using dialer.
+func NewClient(cfg client.Config, params ...roundtrip.ClientParam) (*Client, error) {
+	// Copy cfg.TLS.NextProtos for http TLS, before CheckAndSetDefaults overwrites it for http2.
+	httpNextProtos := cfg.TLS.NextProtos
```

Comment:
How about something like the following?
```go
func cloneHTTPTLSConfig(config *tls.Config) *tls.Config {
    tlsConfig := config.Clone()

    // Clone the tls.Config and set the next protocol. This is needed due to the
    // Auth Server using a multiplexer for protocol detection. Unless next
    // protocol is specified it will attempt to upgrade to HTTP2 and at that point
    // there is no way to distinguish between HTTP2/JSON or GPRC.
    tlsConfig.NextProtos = []string{teleport.HTTPNextProtoTLS}
    
    // This logic is necessary for the HTTP client to force client to always send
    // certificate regardless of the server setting, otherwise client may pick
    // not to send the client certificate by looking at certificate request.
    if len(tlsConfig.Certificates) != 0 {
        cert := tlsConfig.Certificates[0]
        tlsConfig.Certificates = nil
        tlsConfig.GetClientCertificate = func(_ *tls.CertificateRequestInfo) (*tls.Certificate, error) {
            return &cert, nil
        }
    }

    return tlsConfig
}
```